graph2tac/loader/data_server.py

```diff
@@ -171,21 +172,95 @@ def cluster_to_graph(self, cluster):
             ], dtype = self._def_name_dtype)
         )
 
-    
+TRAIN = 0
+VALID = 1
+HOLDOUT = 2 # maybe unnecessary
+
+class Splitter:
+    def lemma(self, d : Definition) -> int:
+        raise Exception("Not implemented")
+    def definition_cluster(self, d : list[Definition]) -> int:
+        raise Exception("Not implemented")
+    def assign_data_server(self, data_server):
+        pass
+
+class SplitByHash(Splitter):
+    def __init__(self, proportions : list[int], random_seed : int):
+        self.proportions = proportions
+        self.random_seed = random_seed
+    def lemma(self, d : Definition) -> int:
+        # to make it identical to vasily's loader
+        ident_64 = np.array(d.node.identity, dtype = np.uint64).item()
+        return get_split_label(ident_64, self.proportions, self.random_seed)
+    def definition_cluster(self, d : list[Definition]) -> int:
+        return 0
+
+class SplitByFilePrefix(Splitter):
+    def __init__(self, prefixes_per_label : list[list[str]]):
+        self.prefixes_per_label = [
+            (label+1, prefixes)
+            for label, prefixes in enumerate(prefixes_per_label)
+        ]
+    def lemma(self, d : Definition):
+        return self.graphid_to_label[d.node.graph]
+    def definition_cluster(self, d : list[Definition]):
+        return self.graphid_to_label[d[0].node.graph]
+    def assign_data_server(self, data_server):
+
+        # conversion from graph indices to filenames
+        graphid_to_fname = {
+            file_data.graph : str(fname)
+            for fname, file_data in data_server._data.items()
+        }
+        assert set(graphid_to_fname.keys()) == set(range(len(data_server._data)))
+        graphid_to_fname = [
+            graphid_to_fname[i]
+            for i in range(len(data_server._data))
+        ]
+
+        # conversion from graph indices to labels + check that all prefixes are used
+        unused_prefixes = set().union(*[prefixes for label,prefixes in self.prefixes_per_label])
+        graphid_to_label = []
+        for fname in graphid_to_fname:
+            label = 0
```

Comment:
Can we also change this 0 to TRAIN?

Reply:
done

Reply:
Sorry, can you also look at all the other `0`'s and `1`'s in this data_server.py and train.py and see if they should be TRAIN or VALID?  Sorry to be so pedantic.

Reply:
I think I did in the last commit, or have I missed some other?